File: managed/node-agent/resources/ynp/executor.py
from commands.provision_command import ProvisionCommand


class Executor:
    def __init__(self, config, args):
        self.config = config
        self.args = args
        self.commands = {
            "provision": ProvisionCommand
        }

    def exec(self):
        command_class = self.commands.get(self.args.command)
        if not command_class:
            raise ValueError(f"Unsupported command: {self.args.command}")
        command_instance = command_class(self.config)

        # Need to validate only in case of onprem nodes.
        if not self.args.extra_vars:
            command_instance.validate()

        if self.args.dry_run:
            command_instance.dry_run()
        elif self.args.preflight_check:
            command_instance.run_preflight_checks()
        else:
            command_instance.execute(self.args.specific_module)

File: managed/node-agent/ynp/executor.go
// Copyright (c) YugabyteDB, Inc.

package ynp

import (
	"context"
	"fmt"
	"node-agent/ynp/command"
	"node-agent/ynp/config"
)

type Executor struct {
	ctx       context.Context
	iniConfig *config.INIConfig
	args      config.Args
	commands  map[string]config.CommandFactory
}

func NewExecutor(iniConfig *config.INIConfig, args config.Args) *Executor {
	return &Executor{
		iniConfig: iniConfig,
		args:      args,
		commands: map[string]config.CommandFactory{
			"provision": command.NewProvisionCommand,
		},
	}
}

func (e *Executor) RegisterCommandFactory(name string, factory config.CommandFactory) {
	e.commands[name] = factory
}

func (e *Executor) Exec(ctx context.Context) error {
	factory, ok := e.commands[e.args.Command]
	if !ok {
		return fmt.Errorf("Unsupported command: %s", e.args.Command)
	}
	command := factory(ctx, e.iniConfig, e.args)
	err := command.Init()
	if err != nil {
		return fmt.Errorf("Failed to initialize %s command: %v", e.args.Command, err)
	}
	// Need to validate only in case of onprem manual.
	if !config.GetBool(e.iniConfig.DefaultSectionValue(), "is_cloud", false) {
		if err := command.Validate(); err != nil {
			return err
		}
	}
	if e.args.ListModules {
		return command.ListModules()
	}
	if e.args.DryRun {
		return command.DryRun()
	}
	if e.args.PreflightCheck {
		return command.RunPreflightChecks()
	}
	return command.Execute()
}
